src/util-misc.c

```diff
@@ -231,6 +231,151 @@ void ShortenString(const char *input,
+    char *endptr = NULL;
+    unsigned long long base = strtoull(p, &endptr, 10);
+    if (errno == ERANGE) {
+        goto overflow;
+    }
+    if (endptr == p) {
+        goto formaterr; // no digits
+    }
+
+    while (isspace((unsigned char)*endptr))
+        endptr++;
+    char unitbuf[16];
+    size_t ulen = 0;
+    const char *up = endptr; // unit processing pointer
+    while (*up && isalpha((unsigned char)*up) && ulen + 1 < sizeof(unitbuf)) {
+        unitbuf[ulen++] = (char)tolower((unsigned char)*up);
+        up++;
+    }
+    unitbuf[ulen] = '\0';
+    /* Any non-space trailing characters invalidate */
+    while (isspace((unsigned char)*up))
+        up++;
+    if (*up != '\0') {
+        goto formaterr;
+    }
+
+    uint64_t multiplier = 1ULL;
+    if (ulen != 0) {
+        /* Match tokens by length then content to be efficient */
+        switch (ulen) {
+            case 1:
+                if (unitbuf[0] == 's') {
+                    multiplier = 1ULL;
+                } else if (unitbuf[0] == 'm') {
+                    multiplier = 60ULL;
+                } else if (unitbuf[0] == 'h') {
+                    multiplier = 3600ULL;
+                } else if (unitbuf[0] == 'd') {
+                    multiplier = 86400ULL;
+                } else if (unitbuf[0] == 'w') {
+                    multiplier = 604800ULL;
+                } else if (unitbuf[0] == 'y') {
+                    multiplier = 31536000ULL;
+                } else {
+                    goto formaterr;
+                }
+                break;
+            case 2:
+                if (memcmp(unitbuf, "hr", 2) == 0)
+                    multiplier = 3600ULL;
+                else
+                    goto formaterr;
+                break;
+            case 3:
+                if (memcmp(unitbuf, "sec", 3) == 0)
+                    multiplier = 1ULL;
+                else if (memcmp(unitbuf, "min", 3) == 0)
+                    multiplier = 60ULL;
+                else if (memcmp(unitbuf, "hrs", 3) == 0)
+                    multiplier = 3600ULL;
+                else if (memcmp(unitbuf, "day", 3) == 0)
+                    multiplier = 86400ULL;
+                else if (memcmp(unitbuf, "wks", 3) == 0)
+                    multiplier = 604800ULL;
+                else if (memcmp(unitbuf, "yrs", 3) == 0)
+                    multiplier = 31536000ULL;
+                else
+                    goto formaterr;
+                break;
+            case 4:
+                if (memcmp(unitbuf, "secs", 4) == 0)
+                    multiplier = 1ULL;
+                else if (memcmp(unitbuf, "hour", 4) == 0)
+                    multiplier = 3600ULL;
+                else if (memcmp(unitbuf, "days", 4) == 0)
+                    multiplier = 86400ULL;
+                else if (memcmp(unitbuf, "week", 4) == 0)
+                    multiplier = 604800ULL;
+                else if (memcmp(unitbuf, "year", 4) == 0)
+                    multiplier = 31536000ULL;
```

Comment:
I always avoid  years in simple parsers, as they're not always the same. Leap years and leap seconds. Presumably why you avoided months.